### PR TITLE
3 i 4. Amelia Sroczyńska

### PR DESCRIPTION
**Pull request by AmeliaSroczynska**
Wybaczcie mocne spóźnienie, umierałam po usunięciu ósemki

### CODE REVIEW

**Comment by Batory12** [comment]
Rzetelnie zrobione zadanie, zbadanych dużo wariantów. Plus za wrzucenie własnej cyfry. Wyniki modeli też bardzo dobre. Jedyna może sugestia, że batch size też może wpłynąć na jakość modelu. Chyba z tych co sprawdzałem to ta praca najlepsza :)

**Comment by AmeliaSroczynska** [comment]
Dodałam jeszcze 4 zadanie, więc ten PR zawiera dwa zadania

**Review by gilian10 — APPROVED** [review]
3 zadanie zrobione świetnie, dużo komentarzy, całość ułożona w logiczną całość, przez co pracę się przyjemnie czyta. Dużo ciekawych eksperymentów i dobra struktura kodu, która umożliwiła szybkie przeprowadzenie testów - troche zaskoczyło mnie odwoływanie się wewnatrz funkcji do modelu jako zmiennej globalnej, a nie przkeazywanej jako argument. Bardzo fajny test na koniec z autorskim obrazkiem.

4 zadanie ok, fajnie, że próbowałaś z innym modelem niż był na zajęciach.

**Review by nuklearbanana — APPROVED** [review]
Zadanie trzecie - obszerne opisy i fajna analiza. 
Fajnie, że testowałaś różne sieci konwolucyjne i testowałaś jak kolejne warstwy oraz ich różne wielkości wpływają na ostateczny wynik. 
Pod koniec świetny pomysł na przetestowanie autorskiego obrazka - oryginalny pomysł.

Odnośnie zadania czwartego model odpowiada na temat, obszernie i w miarę sensownie. 
Można było dodać coś więcej.

**Review by M4jonez — COMMENTED** [review]
Zadanie 4 - Wszystko fajnie działa, jedyne do czego mogę się przyczepić to printowanie tylu fragmentów pociętych dokumentów - pokazanie kilku by wystarczyło i nie trzeba by było niepotrzebnie scrollować połowy notatnika

**Comment by PilarToZiomal** [comment]
**Zad 3**

1. Nie używaj zbioru testowego do monitorowania treningu. Ewaluację w pętli epok wykonuj na walidacji; test uruchamiaj wyłącznie raz na końcu, po doborze hiperparametrów.
2. Bias przy wyborze połowy danych. Tworzysz subsety jako „pierwsze N” próbek, co może wprowadzać systematyczne odchylenie. Zamiast tego wylosuj indeksy (z ustalonym seedem).
3. Walidacja niewykorzystana. Tworzysz `val_baseline`, ale go nie używasz — patrz punkt 1.
4. Wybieraj najlepszy model na podstawie walidacji (best epoch), a nie stanu z ostatniej epoki. Zapisuj `best_state_dict` wg `val_acc` lub `val_loss` i raportuj „best val acc” dla każdego wariantu.
5. Brakuje dodatkowych wizualizacji: krzywych loss/accuracy w funkcji epoki oraz macierzy pomyłek również w wersji znormalizowanej (ułatwia analizę i wyciąganie wniosków).

Bardzo dobra praca. **Zaliczone**.

**Comment by Batory12** [comment]
## Do zad4:
- Załadował Ci się tylko ostatni artykuł, bo nadpisujesz pages
- Takie już czepianie się, ale duplikacja kodu
- Z qwenem moze by trzeba było na gpu spróbować, pytanie jak duży się wybierze. 

Bazowe zadanie dobrze zrobione 👍

**Comment by sonervous7** [comment]
Zad 4.
1. Do wczytywania plików lepiej użyć biblioteki os, albo np. `loader=DirectoryLoader('DANEPDF', glob="**/*.pdf", loader_cls=PyPDFLoader)`
2. Nie trzeba wszystkiego drukować, można kilka, kilkanaście pierwszych chunków, raw format notatnika pokazuje wszystkie outputy i nieładnie to wygląda
3. W porządku model do embeddingów, jedynie co można poprawić ew. to dodanie `normalize_embeddings=True`
4. Większa wersja modelu z cheat_sheet jak najbardziej na plus.
5. Inicjalizacja modelu - tutaj parę drobnych uwag, sugestii ulepszeń:
- Flan-T5 jeśli dobrze kojarzę ma limit tokenów wejściowych 512, więc jeśli retriever zwróci zbyt wiele chunków, nadmiar zostanie ucięty (`truncation=True`). Dobrym pomysłem jest sprawdzić tokenizerem, ile tokenów zajmuje pytanie + kontekst (k chunków) i dostosować parametry (k, długość chunków, max_length), aby uniknąć przypadkowego obcinania treści wejściowej albo zbyt krótkich odpowiedzi. Dodatkowo można zatosować jakiś sposób rerankingu prostego za nim wrzucimy to do modelu. Np. brać te które przekraczają jakiś próg prawdopodobieństwa, cosine_similarity/score. Dodatkowo domyślnie, z tego co sprawdziłem pipeline z "text2text-generation" działa w takim trybie -> `do_sample=False`, czyli deterministycznie, powtarzalnie, ale zawsze najbardziej trafnie, mimo to warto wiedzieć, że do_sample, i temperature, to również parametry wplywające na działanie modelu.

Z tego co widziałem, nie robiłaś zadań dodatkowych, nie mniej jednak praca całkowicie poprawna, a odpowiedzi modelu zadawalające.
